### ISSUE TITLE
Promote Package does not fail when it should

### ISSUE THREAD

**Issue by alesremta-ext43767** [flxbl-io/sfp#53]
**Describe the bug**
Promoting a package which is being built on a MAJOR.MINOR.PATCH combination which has previously been promoted DOES NOT FAIL, instead it prints: "VERSION_ID is already promoted". This is not true, the version number being built is NOT promoted, it is a previous build of the same package.

This hides the fact that a version bump was not performed and only manifests itself when attempting to release to PROD.

packages/core/src/package/promote/PromoteUnlockedPackageImpl.ts
```
try {
    result = await packageVersion.promote();
    result.id = packageVersionData.SubscriberPackageVersionId;
} catch (e) {
    if (e.message.includes('previously released')) {
        SFPLogger.log(`Package ${this.package_version_id} is already promoted, Ignoring`);
    } else throw e;
}
```

**To Reproduce**
- Build and Promote a package
- Do not add any changes in version numbers
- Build and Promote a package again

**Expected behavior**
Error that package cannot be promoted, becuase another build of the same MAJOR.MINOR.PATCH has already been promoted
CLI Error: "You already promoted a package version with this major.minor.patch version number. For a given major.minor.patch number, you can promote only one version."

**Actual behaviour**
Package 04tXXXXXXXX is already promoted*** Ignoring

**Platform Details (please complete the following information):**
Tested with older version: sfp  -- The DX@Scale CLI -Version:25.5.3 -Release:December 23
But checked code in new SFP repository and it is still the same.
 - OS: 
 - Version [e.g. CLI Version  eg: 1.6.6]
 - Salesforce CLI(sfdx cli) Version:
 - CI Platform: GitHub Actions

**Additional context**

**Comment by dieffrei** [comment]
I agree, I felt that pain many times. I will take a look @alesremta-ext43767 what we can do

**Comment by azlam-abdulsalam** [comment]
We need to account for retries while using release command.  An inability to promote should not hinder one to retry a release.  I think the best option is to do a dry run and check for issues upfront. For eg in sfops, 

<img width="982" alt="image" src="https://github.com/flxbl-io/sfp/assets/43767972/696ffe3d-e181-4810-b94f-23b7ddf7a47f">

**Comment by alesremta-ext43767** [comment]
Hi @azlam-abdulsalam I think it should be fine so long as there is a distinction between a) promoting already promoted version again (SF CLI will just give a "success" response) and b) promoting a different build of an already promoted major.minor.combo (SF CLI will throw an error)

**Comment by odaybov-cen56981** [comment]
hi @azlam-abdulsalam I'm also involved to this topic. just i vote for the ver **b) promoting a different build of an already promoted major.minor.path combo (SF CLI will throw an error)**

- thx.

**Comment by dieffrei** [comment]
I agree with @azlam-abdulsalam there are some edge cases that it could be tricky.
However, we can have a flag for that "--fail-if-promoted"
@odaybov-cen56981 @alesremta-ext43767 @azlam-abdulsalam

**Comment by azlam-abdulsalam** [comment]
The need is mainly driven by the 'promoteBeforeDeployment' attribute in release defn, where most orgs just promote it before deploying to prod.  I think @dieffrei's solution could be perfect, we could add an additional attribute or flag

@alesremta-ext43767 is this the promote command or release command?

**Comment by alesremta-ext43767** [comment]
We are using the "promote command". But to be honest, for me this is much more low level. Whatever you do, promote or release, ultimately does this:
`sf package version promote --package xxxxx --target-devhub-org xxxx`
And my issues is that _when_ this command fails, the SFP command must fail. Because it means that the promotion could not be done and therefore the package cannot be installed in PROD. Right now this is not the case. SFP catches all "errors" and assumes this was due to the fact that the given Package Version is already promoted. But that is not always the case. Sometimes it's not the specific Package Version that was already promoted but rather a different Package Version with the same MAJOR.MINOR.PATCH combo (just different build, from previous release definition perhaps) and then this _cannot_ be ignored.

**Comment by alesremta-ext43767** [comment]
@dieffrei @azlam-abdulsalam I would say it's a lot more important for the Promote command. For release, if this issue is in place, you're going to fail right after (attempting to install). But if your process is to prepare artefacts ahead of time to do a just in time release (perhaps in a short prod maintenance window) you want to be as close to 100% sure it'll install successfully as you can. So you would promote ahead of time and rely in the fact that this package _is_ promoted successfully.

**Comment by dieffrei** [comment]
@alesremta-ext43767 In terms of effort, the release command uses the promote command. 
We will change both anyway.
Probably, I will add the flag for both commands
